#31: Add "Home Office" field in "Worklog" doctype

amatyaPratik opened the pull request:
Hello @marius-meissner ,

I have implemented the feature.
It is ready for review.

### Changelog

- Added `is_home_office` field to **_Worklog_** doctype
- Implemented _**localStorage**_ persistence for home office preference
- UX enhancements:
  - Worklog form now **_auto-populates_** **_Employee_** and **_Log Time_** fields on load
  - Added form validation for required fields
  - placeholder texts in fields added/updated as helpful hints
  - '**_Resume Work_**' status rendered in UI after submission of Checkin type of **_Break_**


### Expected UI:

![Screenshot_20-8-2025_193819_erp neomium local](https://github.com/user-attachments/assets/f3bfe1fe-7f7c-4c00-8488-4fe782241b66)

<img width="349" height="593" alt="image" src="https://github.com/user-attachments/assets/cb5a8e00-3172-43a8-9e0b-b1fdb23911fc" />

Thank you

amatyaPratik committed:
#27 add meaningful Checkin status for work resumption after Break

amatyaPratik committed:
#31 Added 'is_home_office' field to Worklog doctype, enhanced frontend implementation (preference storage, validation, UI)

marius-meissner commented on hr_time/translations/de.csv:
```diff
@@ -97,8 +101,17 @@ Unknown action provided,Aktion unbekannt
 Describe your task here,Beschreiben Sie hier Ihre Aufgabe
+Summarize your task here,Fassen Sie Ihre Aufgabe hier zusammen.
 External reference,Externe reference
+Home Office,Heimbüro
```
The term “home office” has now become well established in German and, in my opinion, no longer needs to be translated.

amatyaPratik replied:
Understood.

marius-meissner commented on hr_time/hr_time_management/doctype/worklog/worklog.js:
```diff
@@ -1,31 +1,161 @@
+                const prevWfhPref = HR_TIME?.LS_KEYS?.PREV_WFH_PREF || 'neo_hr_time_last_wfh_value'
+                localStorage.setItem(prevWfhPref, frm.doc.is_home_office.toString());
+            }
+        }catch (error){
```
```suggestion
        } catch (error) {
```

marius-meissner commented on hr_time/hr_time_management/doctype/worklog/worklog.js:
```diff
@@ -1,31 +1,161 @@
+
+    /** Auto set employee field by clicking the first option in the Employee select field */
+    attemptAutoSelectEmployee: function(employeeField, frm) {
+        if(frm.doc.employee) return;
+
+        const wrapper = employeeField.wrapper;
+        let attempts = 0;
+        const maxAttempts = 30; // Prevent infinite looping
+
+        const attemptAutoSelect = setInterval(() => {
+            attempts++;
+            const firstOption = wrapper.querySelector('ul div:nth-child(1) p');
+            const empNameSpan = firstOption?.querySelector('span.small');
+            const empTitle = firstOption?.getAttribute('title');
+
+            if (firstOption && empNameSpan && empTitle) {
+                clearInterval(attemptAutoSelect);
+                frm.set_value('employee', empTitle);
+
+                const helpBox = wrapper.querySelector('p.help-box');
+                if (helpBox) helpBox.textContent = empNameSpan.textContent;
+                
+                const btnNow =  document.querySelector(".datepicker--button[data-action='today']")
+                if(btnNow) btnNow.click();
+                
+                const taskDescEditor = frm.fields_dict.task_desc?.$wrapper
+                    .get(0)
+                    .querySelector('.ql-editor');
+
+                if (taskDescEditor) {
+                    taskDescEditor.focus();
                 }
-            });
+            } else if (attempts >= maxAttempts){
+                clearInterval(attemptAutoSelect)
+            }
+        }, 200); // Check if Options loaded every 200ms
+    },
+};
```
Instead of clicking UI elements, please use the proper APIs to fetch the values directly.
This is more stable, less complex, and reduces the risk of breaking changes in the future (if the UI classes change).

E.g.
```JS
frappe.call({
  method: "hr_time.api.employee.api.get_current_employee_id",
  callback: (r) => r.message && frm.set_value('employee', r.message),
});
```

amatyaPratik replied:
Thank you for the feedback.

This came into my mind as well considering the said _"changes in future_" aspect.
My thinking here was to reduce reduce backend calls and simply just click the buttons via code like what the user already is supposed to do manually.

But the concern is valid.
I will implement it via API calls like you suggested.

marius-meissner reviewed (changes requested):
Changes have been deployed to staging.
Seems to work good.
@TobiNeo Please have a look from your side as well

TobiNeo commented:
> Changes have been deployed to staging. Seems to work good. @TobiNeo Please have a look from your side as well

Hi @amatyaPratik,

I just took a look at staging and seems to be working nicely, the dropdown for choosing yes/no makes sense. Now, one thing which I noticed:
<img width="2616" height="1319" alt="grafik" src="https://github.com/user-attachments/assets/af30d5da-727c-451d-8e94-021bf8843672" />

When you click on "enter complete detail" in the bottom, whatever has been entered before is gone and the employee in the worklog page was automatically set to HR-EMP-00001. Now it could be that this is just a staging system issue, and generally the function is probably not needed as all details can already be entered in the popup window. If it is easy to fix feel free - otherwise the "enter complete detail" link could also just be omitted? What do you think @marius-meissner?

Cheers
Tobi

amatyaPratik commented:
>>
> When you click on "enter complete detail" in the bottom, whatever has been entered before is gone and the employee in the worklog page was automatically set to HR-EMP-00001. Now it could be that this is just a staging system issue, and generally the function is probably not needed as all details can already be entered in the popup window. If it is easy to fix feel free - otherwise the "enter complete detail" link could also just be omitted?
>>

Hello @TobiNeo ,

Thank you for noticing this and for the feedback.

This is true.
However, it is not a bug/issue, just an unimplemented feature.

As of now (some questions):

1. after entering the **_"Task Description"_** in the Worklog Popup and then clicking the _"enter complete detail"_ link, the entered text hasn't been implemented to be preserved and carried to the clean Worklog document form. _(But this should not be very complicated to implement. Personally, I think this button does hold a value for the user and improving UX.)_

2. The auto populating of Employee was designed for **_non-admin /employee_** use cases. For admins, it simply selects the first employee from the options list. I thought this would be acceptable as I understood, admins accounts are not supposed to fill worklog for themselves but can fill it for other employees ? _(Pls correct me if I am wrong. I understood admins to only be involved in administration and not fill worklogs. Perhaps I am missing a piece here ? )_


I will proceed with the implementation of the suggested changes if it is alright

TobiNeo commented:
Hi @amatyaPratik, 
Sounds good - auto populating only for employees is fine. If you think that the button does hold value, feel free to implement the data to be carried over

Thanks,
Tobias

amatyaPratik commented:
Thank you for clarifying. I was just curious about the default **_Administrator_** role (just to be sure).
And now my confusions are clear.

Regarding employee with managerial roles, the API approach will also fix this "_select first in the list_" issue we are seeing 👍.

The code is almost complete at the moment.
I will notify with updates once it is done.

Thank you.

amatyaPratik committed:
#31 Implemented api based Employee and Log time field auto-fill, autofocus on empty field, code enhancements

amatyaPratik commented:
Good afternoon @marius-meissner ,

I have implemented the suggested changes.
It is ready for review.

### Changelog

- Added api to get current **_Employee_**'s detail
- Changed autofilling of **_Employee_** and **_Log Time_** fields to API based _(rather than UI triggered)_
- **_Easy Checkin Dialog_**'s Worklog form now carries over filled data to the **_Worklog_** document entry page. 
- Code refactoring
- UX enhancements:
  - Worklog now autofocuses on the first empty field _(rather than always on **_Task description_**)_ 
  
  
Please let me know if there are further issues.
  
Thank you

marius-meissner reviewed (approved):
Implement API backend + frontend calls are looking good and clean :+1: 
Changes have been deployed to staging and from my side looks good so far.

@TobiNeo Please test from your side as well.

TobiNeo commented:
Hi,
Just had a look - looks good from my side as well, nice UX. Carrying data to the full form also seems to work well.

Thanks 👍
